### PR TITLE
ScikitLearn.jl wrap

### PR DESCRIPTION
**Pull request by cstjean**
`get-params` wasn't working out with the current type definition. I've decided to wrap `GLRM` instead, it's going to be more consistent with the rest of the ScikitLearn.jl algorithms, and allows me to expose more hyperparameters.

I branched from dataframe-ux this time.

PCA works fine, but I'm having issues with K-Means. I've translated [this example](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_digits.html), and instead of getting a nice Voronoi diagram, I get this:

screen shot 2016-04-28 at 14 05 14

Clearly not random, but not quite Voronoi either! To get this picture, I call `transform(kmeans, X)` where X is a two-column matrix containing all (x,y) coordinates. `transform` uses `FixedLatentFeaturesConstraint`. Is there any reason why this would be incompatible with K-Means? Its output doesn't look right, even with abs_tol=1.e-20 (it converges in ~20 iterations)

```
2430x10 Array{Float64,2}:
  0.0       0.0        0.0      …  0.0      0.0      0.0      1.0     
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
 -0.570026  0.714604  -2.94676     2.10088  2.1646  -1.42798  0.191742
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
  0.0       0.0        0.0      …  0.0      0.0      0.0      0.0     
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
  0.0       0.0        0.0         0.0      0.0      0.0      1.0     
  0.0       0.0        0.0         0.0      0.0      0.0      0.0     
...
```

(10 columns because there are 10 digits). Those 10 rows correspond to neighboring pixels. They should all be classified in the same cluster. But even ignoring the anomaly on row 4, the other rows are not consistent at all.

It would be straight-forward to implement `transform` using the distance from each centroid, but if there's something wrong with `transform`, it's better to fix that instead. Any ideas?

### CODE REVIEW

**Comment by cstjean** [comment]
I did some more testing with dataframe-ux: 

```julia
srand(21)
# Create two very disjoint Gaussians at (-10, -10) and (5,5)
gaussian1 = randn(100, 2) + 5
gaussian2 = randn(50, 2) - 10
A = vcat(gaussian1, gaussian2)

km = kmeans(A, 2)
init_kmeanspp!(km)
X, Y, ch = fit!(km)
```

Y looks fine:

```
2x2 Array{Float64,2}:
  4.86053   4.87729
 -9.98556  -9.84932
```

X has some issues, but isn't too bad. Are the unity constraints supposed to be "soft"?

```julia
> sort(X'[1:100, 1])

100-element Array{Float64,1}:
 -1.06409     # should all be 1.0
 -0.581968
 -0.574768
  1.0     
  1.0     
  1.0     
  1.0     
  1.0     
...
```

If I use `FixedLatentFeaturesConstraint`, it looks quite a bit worse.

```julia
srand(3)
ry_fixed = [FixedLatentFeaturesConstraint(km.Y[:, i])
            for i=1:size(km.Y, 2)]
km_fixed = GLRM(A, km.losses, km.rx, ry_fixed, km.k)
X2, _, ch = fit!(km_fixed, ProxGradParams(abs_tol=1.e-30, max_iter=100000))
sum(X2', 1)
```
```
Fitting GLRM
Iteration 10: objective value = 1118.1314411075239

1x2 Array{Float64,2}:
 86.6146  37.6278      # should be 100 50
```

I get that GLRM is a very general framework, and asking it to do KMeans in particular may put stress on its assumptions. Or did I do something wrong?

**Comment by madeleineudell** [comment]
Hi Cedric,

To answer your first post: the difficulty is that this package implements
one fitting routine - proximal gradient - for all problems. For the kmeans
problem, this algorithm is *not* the same as Lloyd's algorithm (the
standard kmeans algorithm), and as you notice, it works quite a bit worse.
That's essentially because we do alternating updates, rather than
alternating minimization, at each step.

You can make the algorithm do something more like alternating minimization,
for convex problems using ProxGradParams(inner_iter=10). The higher the
inner iteration count, the better the algorithm should approximate
alternating minimization. But for nonconvex problems, like kmeans, even
this will not be exactly like minimizing the nonconvex objective. I think
it's a pretty interesting question whether there are other simple
algorithms that you can write just using gradients of objectives and proxs
onto constraints that do something more reasonable for nonconvex problems.

To answer your second post: I located a bug in the UnitOneSparseConstraint
evaluate method that broke my linesearch routine. I've fixed it and now all
the unity constraints hold exactly. My guess is that this should make your
first example work better, too. I've pushed the fix onto the dataframe-ux
branch.

Madeleine

On Fri, Apr 29, 2016 at 5:33 AM, Cédric St-Jean <notifications@github.com>
wrote:

> I did some more testing with dataframe-ux:
>
> srand(21)# Create two very disjoint Gaussians at (-10, -10) and (5,5)
> gaussian1 = randn(100, 2) + 5
> gaussian2 = randn(50, 2) - 10
> A = vcat(gaussian1, gaussian2)
>
> km = kmeans(A, 2)init_kmeanspp!(km)
> X, Y, ch = fit!(km)
>
> Y looks fine:
>
> 2x2 Array{Float64,2}:
>   4.86053   4.87729
>  -9.98556  -9.84932
>
> X has some issues, but isn't too bad. Are the unity constraints supposed
> to be "soft"?
>
> > sort(X'[1:100, 1])
> 100-element Array{Float64,1}:
>  -1.06409     # should all be 1.0
>  -0.581968
>  -0.574768
>   1.0
>   1.0
>   1.0
>   1.0
>   1.0     ...
>
> If I use FixedLatentFeaturesConstraint, it looks quite a bit worse.
>
> srand(3)
> ry_fixed = [FixedLatentFeaturesConstraint(km.Y[:, i])
>             for i=1:size(km.Y, 2)]
> km_fixed = GLRM(A, km.losses, km.rx, ry_fixed, km.k)
> X2, _, ch = fit!(km_fixed, ProxGradParams(abs_tol=1.e-30, max_iter=100000))sum(X2', 1)
>
> Fitting GLRM
> Iteration 10: objective value = 1118.1314411075239
>
> 1x2 Array{Float64,2}:
>  86.6146  37.6278      # should be 100 50
>
> I get that GLRM is a very general framework, and asking it to do KMeans in
> particular may put stress on its assumptions. Or did I do something wrong?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/pull/55#issuecomment-215699226>
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
*https://courses2.cit.cornell.edu/mru8
<https://courses2.cit.cornell.edu/mru8>*
(415) 729-4115

**Comment by cstjean** [comment]
Your changes fixed the unity constraint issue, thank you! The results look better, but they still haven't converged with `inner_iter=60`.

screen shot 2016-05-03 at 09 06 04

I rewrote `predict` using euclidean distance (via Distances.jl), and of course that solved the issue...

In any case, users of ScikitLearn.jl will come looking for PCA and KMeans in the ScikitLearn.jl model page, and maybe it would be best to get these from MultivariateStats.jl and Clustering.jl, while using LowRankModels for more complex problems that take advantage of its generality? Have you benchmarked LowRankModels' NMF, QPCA and RPCA against standard algorithms?

**Comment by madeleineudell** [comment]
Cedric, thanks for your thorough investigation; ScikitLearn will be much
the stronger for it.

My focus in writing LowRankModels was on being able to embed heterogeneous
(boolean, real, ordinal, etc) data sets, and to switch easily between
different losses and regularizers. The fact that it's possible to write a
single algorithm (alternating proximal gradient) that runs on all these
problems is a curiosity, but certainly does not lead to the best method for
all problems. I'd be very happy to include other algorithms in the package.
In fact, I'm currently working on a few other algorithms that I am going to
add to the package. Alternating minimization (Lloyd's algorithm for kmeans)
could easily be one of them. But the user would still need to know to
select the right algorithm for their problem.

I did not benchmark the proximal gradient method for NMF, QPCA and RPCA
against standard methods, because building faster software for problems
that already have solvers was not a priority for me. But I can tell you
what I know about which problems these methods will work well for. Whenever
the loss function is convex, and the regularizer is quadratic, then I have
a fair amount of confidence that init_svd! followed by the proximal
gradient method will produce the global optimum with high probability. For
many of these problems, there are proofs that this works with high
probability when your data is generated by a statistical model. For NMF,
the proximal gradient fit method will work ok, but probably won't be as
good as custom NMF solvers.

Let me know if you have any more questions as you decide which routines to
rely on for ScikitLearn. I'm happy to implement alternating minimization
for UnitOneSparseConstraint if it would help, but I'm afraid LowRankModels
will never be a one-stop-shop for the *best* algorithm for every problem it
can model.

On Tue, May 3, 2016 at 6:55 AM, Cédric St-Jean <notifications@github.com>
wrote:

> Your changes fixed the unity constraint issue, thank you! The results look
> better, but they still haven't converged with inner_iter=60.
>
> [image: screen shot 2016-05-03 at 09 06 04]
> <https://cloud.githubusercontent.com/assets/8622776/14984211/9258fb98-110f-11e6-9827-9c3462bf3d9c.png>
>
> I rewrote predict using euclidean distance (via Distances.jl), and of
> course that solved the issue...
>
> In any case, users of ScikitLearn.jl will come looking for PCA and KMeans
> in the ScikitLearn.jl model page, and maybe it would be best to get these
> from MultivariateStats.jl and Clustering.jl, while using LowRankModels for
> more complex problems that take advantage of its generality? Have you
> benchmarked LowRankModels' NMF, QPCA and RPCA against standard algorithms?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/pull/55#issuecomment-216535199>
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
*https://courses2.cit.cornell.edu/mru8
<https://courses2.cit.cornell.edu/mru8>*
(415) 729-4115

**Comment by cstjean** [comment]
Madeleine, 

That helps a lot, thank you. From my side, it's not a lot of work to support PCA, KMeans, etc. We can advertise them inside ScikitLearn.jl, but I'll add a note about efficiency. scikit-learn's PCA and KMeans cannot deal with missing values; that's a strong point in favour of GLRM. When/if better algorithms are implemented in LowRankModels.jl, we can change the defaults for each model. 

> My focus in writing LowRankModels was on being able to embed heterogeneous (boolean, real, ordinal, etc) data sets, and to switch easily between different losses and regularizers.

That's really cool. I should do a better job with the `SkGLRM` interface, especially to allow different regularizers/losses for each column. It would be nice also to use `LowRankModels` for imputation: scikit-learn's [imputer](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Imputer.html) just uses the mean/median/mode along each column.

I'll try to get a first version ready for merging this week, but hopefully we can revisit these issues later this year.

**Comment by cstjean** [comment]
The code is ready for review. If it looks reasonable, I'll add docstrings and a section to the README. I didn't have time to test NMF, QPCA and RPCA to the same extent that I [tested KMeans/PCA](https://github.com/cstjean/ScikitLearn.jl/blob/master/examples/Plot_Kmeans_Digits_Julia.ipynb), but their output looks OK.

NMF's comment in simple_glrms says:

```julia
# nonnegative matrix factorization
# minimize_{X<=0, Y>=0} ||A - XY||^2
```

Shouldn't it be `X>=0`?

For handling missing values, I used `obs`, and it defaults to all non-NaN entries. What's the difference between `observed_examples` and `observed_features`?

Am I correct in assuming that one of rx.scale, ry.scale, loss.scale is redundant? I only used rx_scale and ry_scale.

I used your defaults, but if you feel like other defaults are more appropriate for ScikitLearn.jl users, we should change them.

**Comment by madeleineudell** [comment]
I should be able to review it over the weekend. In the meantime, answers to
your questions are below.

> The code is ready for review. If it looks reasonable, I'll add docstrings
> and a section to the README. I didn't have time to test NMF, QPCA and RPCA
> to the same extent that I tested KMeans/PCA
> <https://github.com/cstjean/ScikitLearn.jl/blob/master/examples/Plot_Kmeans_Digits_Julia.ipynb>,
> but their output looks OK.
>
> NMF's comment in simple_glrms says:
>
> # nonnegative matrix factorization# minimize_{X<=0, Y>=0} ||A - XY||^2
>
> Shouldn't it be X>=0?
>
Yes, it should. I just pushed a change to master.

> For handling missing values, I used obs, and it defaults to all non-NaN
> entries. What's the difference between observed_examples and
> observed_features?
>
`observed examples` is a list of length n giving the rows observed for
every column; `observed_features` is a list of length m giving hte columns
observed for every row. It's useful to form these ahead of time to run the
proximal gradient fitting algorithm faster. It's find to provide `obs` and
let the package calculate them for you.

> Am I correct in assuming that one of rx.scale, ry.scale, loss.scale is
> redundant? I only used rx_scale and ry_scale.
>
That's true, but less true than it might seem. The loss and regularizer ry
for each column can be independently scaled; so it's really only rx, that
is forced to be the same for every row, that can be discarded with no loss
of model expressiveness.

> I used your defaults, but if you feel like other defaults are more
> appropriate for ScikitLearn.jl users, we should change them.
>
My guess is they're fine...

> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/pull/55#issuecomment-216991857>
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
*https://courses2.cit.cornell.edu/mru8
<https://courses2.cit.cornell.edu/mru8>*
(415) 729-4115

**Comment by cstjean** [comment]
Hey Madeleine, I think this is ready enough for now. Perfect is the enemy of good and all that. Feel free to tweak it to improve consistency with the rest of LowRankModels. Should I rebase/squash it on master?

**Comment by madeleineudell** [comment]
@cstjean, this looks good to me. Go ahead and rebase/squash on master and I'll merge.
